### DISCUSSION TITLE
_CastError (Null check operator used on a null value) in registerBackgroundMessageHandler could be improved

### DISCUSSION THREAD

**Discussion by hecklatcipsoft** [firebase/flutterfire#5636] [Feature request]
## Bug report

**Describe the bug**
In File firebase_messaging_platform_interface-2.0.0\lib\src\method_channel\method_channel_messaging.dart
in function registerBackgroundMessageHandler 

bgHandle and userHandle are assigned using 
```
 final CallbackHandle bgHandle = PluginUtilities.getCallbackHandle(
          _firebaseMessagingCallbackDispatcher)!;
      final CallbackHandle userHandle =
          PluginUtilities.getCallbackHandle(handler)!;
```
But PluginUtilities.getCallbackHandle can return null, which will then result in an exception:

Exception has occurred.
_CastError (Null check operator used on a null value)

### Steps to reproduce

1. Clone [this repository](https://github.com/hecklatcipsoft/firebase_exception_minimal.git)
2. Run the app
3. See error 

### Expected behavior
No Exception should occur.

---

## Additional context

---

### Flutter doctor

Run `flutter doctor` and paste the output below:

<details><summary>Click To Expand</summary>

```
PS C:\Development\github\hecklatcipsoft\firebase_exception_minimal> flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 2.0.2, on Microsoft Windows [Version 10.0.19041.804], locale en-GB)
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.1.0)
[✓] VS Code (version 1.54.3)
[✓] Connected device (3 available)

• No issues found!
```

</details>

---

### Flutter dependencies

Run `flutter pub deps -- --style=compact` and paste the output below:

<details><summary>Click To Expand</summary>

```
PS C:\Development\github\hecklatcipsoft\firebase_exception_minimal> flutter pub deps -- --style=compact
Dart SDK 2.12.1
Flutter SDK 2.0.2
firebase_exception_minimal 1.0.0+1

dependencies:
- cupertino_icons 1.0.2
- firebase_core 1.0.1 [firebase_core_platform_interface firebase_core_web flutter meta]
- firebase_messaging 9.0.0 [firebase_core firebase_core_platform_interface firebase_messaging_platform_interface firebase_messaging_web flutter meta]
- flutter 0.0.0 [characters collection meta typed_data vector_math sky_engine]

dev dependencies:
- flutter_test 0.0.0 [flutter test_api path fake_async clock stack_trace vector_math async boolean_selector characters charcode collection matcher meta source_span stream_channel string_scanner term_glyph typed_data]

transitive dependencies:
- async 2.5.0 [collection]
- boolean_selector 2.1.0 [source_span string_scanner]
- characters 1.1.0
- charcode 1.2.0
- clock 1.1.0
- collection 1.15.0
- fake_async 1.2.0 [clock collection]
- firebase_core_platform_interface 4.0.0 [flutter meta plugin_platform_interface]
- firebase_core_web 1.0.1 [firebase_core_platform_interface flutter flutter_web_plugins js meta]
- firebase_messaging_platform_interface 2.0.0 [firebase_core flutter meta plugin_platform_interface]
- firebase_messaging_web 1.0.1 [firebase_core firebase_core_web firebase_messaging_platform_interface flutter flutter_web_plugins js meta]
- flutter_web_plugins 0.0.0 [flutter js characters collection meta typed_data vector_math]
- js 0.6.3
- matcher 0.12.10 [stack_trace]
- meta 1.3.0
- path 1.8.0
- plugin_platform_interface 2.0.0 [meta]
- sky_engine 0.0.99
- source_span 1.8.0 [charcode collection path term_glyph]
- stack_trace 1.10.0 [path]
- stream_channel 2.1.0 [async]
- string_scanner 1.1.0 [charcode source_span]
- term_glyph 1.2.0
- test_api 0.2.19 [async boolean_selector collection meta path source_span stack_trace stream_channel string_scanner term_glyph matcher]
- typed_data 1.3.0 [collection]
- vector_math 2.1.0
```

</details>

---

**Comment by markusaksli-nc** [comment]
Hi @hecklatcipsoft 
The callback handle cannot be null, otherwise, the method can't be registered for the background messaging handler.
You need to provide a top-level function that is not anonymous.
```dart
void _bootStrapFirebase() async {
  await Firebase.initializeApp();
  FirebaseMessaging.onBackgroundMessage(onBackgroundMessageHandler);
}

Future<void> onBackgroundMessageHandler(RemoteMessage message) async {
  print("Handling a background message ${message.data}");
  //You need to do everything in here
  //If you need to do anything with firebase you also need to call
  //await Firebase.initializeApp();
}
```
Thank you

**Comment by hecklatcipsoft** [comment]
Hi @markusaksli-nc,

Thanks for your answer, I didn't read the comment that clearly states that it has to be a top level function.

However you still might want to update the comment because there also isn't an ArgumentError we could handle to deal with a problem.

The async registration just fails and the log is lost in the debug console.

Changing registerBackgroundMessageHandler to not be async might help mitigate the problem because then at least some exception hits the consumer.

```
@override
  void registerBackgroundMessageHandler(BackgroundMessageHandler handler) {
    if (defaultTargetPlatform != TargetPlatform.android) {
      return;
    }

    if (!_bgHandlerInitialized) {
      _bgHandlerInitialized = true;
      final CallbackHandle bgHandle = PluginUtilities.getCallbackHandle(
          _firebaseMessagingCallbackDispatcher)!;
      final CallbackHandle userHandle =
          PluginUtilities.getCallbackHandle(handler)!;

      channel.invokeMapMethod('Messaging#startBackgroundIsolate', {
        'pluginCallbackHandle': bgHandle.toRawHandle(),
        'userCallbackHandle': userHandle.toRawHandle(),
      });
    }
  }

```

**Comment by markusaksli-nc** [comment]
I'm pretty sure MethodChannels are strictly async so you cannot delegate synchronous calls in Flutter right now so I'm not sure that would be possible. Though maybe there could be better feedback here.

**Comment by tonnyavery** [comment]
I can agree that this message is very confusing.
Providing an async named method that is not top-level function just results in a `(Null check operator used on a null value)` error that doesn't help much.

Also, the whole concept is a bit unclear. This callback is needed to be able to show local notifications somehow. It's not possible to pass any custom arguments into the callback, how are we supposed to show custom notifications?

**Comment by JabalSoft** [comment]
We have spent much time trying to upgrade our packages to the latest version after upgrading flutter to V2.0.0. 

This issue really affects our upgrades, 

Is there any workaround to solve this issue? 😟

**Comment by tonnyavery** [comment]
@JabalSoft There is no issue. Everything works if you provide a top level named function as a callback because flutter_messaging users isolate internally.
The issue I see is that if you use some other dependency to show notification, you need to find a way to use it from a top level function somehow

**Comment by JabalSoft** [comment]
> @JabalSoft There is no issue. Everything works if you provide a top-level named function as a callback because flutter_messaging users isolate internally.
> The issue I see is that if you use some other dependency to show notification, you need to find a way to use it from a top-level function somehow

Your comment really helps us, we convert our anonymous function to class function but not top-level (outside class). 

After we put our function outside our class OR make it static the issue solved.

Maybe coloring the lines of the instructions on the documentation page with yellow or red is a good idea 😅.

Thanks.

**Comment by markusaksli-nc** [comment]
@AAverin What do you mean by the local notifications bit? That is what `onMessage` is for. The background message handler is meant for handling data-only messages that can be used for various functional purposes (such as phone auth code auto-retrieval in firebase_auth).

**Comment by Salakar** [comment]
I think we should improve the internal typing here so it doesn't throw a cast error at least

**Comment by erayhamurlu** [comment]
add this line  "WidgetsFlutterBinding.ensureInitialized()" before   "await Firebase.initializeApp()" ,

**Reply by harshkumarkhatri** [reply]
It's pretty evident to add it before implementing firebase in the app else nothing would work

**Reply by leeprobert** [reply]
This did work for me 👍

**Comment by choakumndi** [comment]
Hi, please i still don't understand why this error persist in my code...
![image](https://user-images.githubusercontent.com/5228299/137989189-1fd5c1c0-562f-473a-b819-2dc4e93be56f.png)
I am trying to get the doc path but the _CastError keeps showing up. I need help.
